### ISSUE TITLE
Service Worker: allow service worker to checkForUpdate and activateUpdate before bootstraping the AppModule

### ISSUE THREAD

**Issue by romanovma** [angular/angular#39008]
# 🚀 feature request
Provide a way to check for updates and activate them before the main AppModule is bootstrapped.

### Relevant Package
This feature request is for @angular/service-worker

### Description
Currently, if we use the service worker module(which we want to use in order to support offline mode) we have no choice but always serve a cached(potentially outdated) version when a user opens our app. And only then we can check for updates in the background.

What we want is to check for updates before bootstrapping the app and if there are any then first activate them and only after that bootstrap an app. We want to be sure that a user starts her working day with a fresh version of our app (if for example there was an overnight update).

We considered force page reload and showing a prompt for a user to activate a new version but that would not be the best UX choice in our scenarios(we do updates a couple of times daily).

### Describe the solution you'd like
Provide an API to access service worker's `checkForUpdate`, `available` and `activateUpdate` properties outside of a module/service so it can be done in `main.ts`.

### Describe alternatives you've considered
We can't remove the service worker since we want offline support.
We can't force refresh as it's a bad UX for us(we can on navigation between routes but that would be too late)
We can't show prompt as it will appear almost every morning for the user.

We consider showing a spinner in app.component until we make all checks through the ServiceWorkerModule as an alternative but that would be a waste of the device's(in case of mobile) resources to parse and execute all JS to then find out it's not needed.

**Comment by gkalpak** [comment]
Coincidentally, we landed #38565 a few days ago, which adds support for configuring the SW to serve navigation requests from the server (but still fall back to the cache when offline). This should address your issue for the most part. (You might want to apply further configuration changes, such as automatically install available updates (emitted on `SwUpdate#available`), eliminate unhashed resources from your asset-groups, etc.)

It will probably be released in v11 :tada:

##
In the meantime, (or if you this does not cover your needs 100%), there are work-arounds you can use (see below). However, exposing `SwUpdate` outside of the Angular app (i.e. outside of dependency injection, etc.) is not something that aligns with Angular's philosophy, I'm afraid.

Work-arounds:

1. The communication with the SW happens via the native [ServiceWorker][1]'s `postMessage()` method ([source code][2]), so it is easy to implement these yourself.
2. You could instantiate `SwUpdate` yourself (but you need to manually create and pass it an instance of [NgswCommChannel][3]).
3. You could avoid including `index.html` in an [asset group][4] and include it in a [data group][5] configured with the `freshness` [strategy][6] and a long [maxAge][7].

All of the above are work-arounds, of course, and not officially supported or recommended (i.e. use at your own risk :grin:).

##
WDYT, @romanovma? Would #38565 and/or one of the work-arounds cover your needs?
 
[1]: https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker
[2]: https://github.com/angular/angular/blob/323be39297450fa5a4a70f800511722b4964dec2/packages/service-worker/src/low_level.ts#L121-L124
[3]: https://github.com/angular/angular/blob/323be39297450fa5a4a70f800511722b4964dec2/packages/service-worker/src/low_level.ts#L87
[4]: https://angular.io/guide/service-worker-config#assetgroups
[5]: https://angular.io/guide/service-worker-config#datagroups
[6]: https://angular.io/guide/service-worker-config#strategy
[7]: https://angular.io/guide/service-worker-config#maxage

**Comment by romanovma** [comment]
First of all, Wow! thanks a lot for such a quick and detailed response @gkalpak! That's truly awesome support!
The PR you mentioned is exactly what we need! I hope that it lands in v11 and that we are able to upgrade our current app's Angular version accordingly 😃 
Do you consider adding `timeout` option so that it's possible to fall back to cached version in case of low connection?

Workarounds: 
 - For both 1. and 2. I just realized I can do the job within APP_INITIALIZER function which follows Angular philosophy and more or less happens before the bootstrap
 - for 3. I don't think I really understand the complications of the approach.  If moving index.html into a data group really works  in the way I described above (I briefly looked into the source code - it seems to work), and the issue is only with semantics(that I put an `asset` configuration into the `data` configuration section) then I would go that way.

Alternatively as an option 4., I think of checking for `availability` Observable inside APP_INITIALIZER function, and if it emits anything within for instance 1000ms then I'd activate the update and reload the page (we will track over time when this emit happens and will adjust that 1000ms to a proper value according to the collected data).
This approach is based on my understanding that check for updates happens on the navigation event to our app(made by service worker even before we execute the application's code). This means that probably in many cases updates will be already available when the app is being bootstrapped. 

Solution is far from ideal for sure because we will have this 1000ms delay every time we load the page. And we will have this delay twice in a row in case we found an update and reload the page immediately. We could avoid the delay if we have a way to quickly check if an update is even detected but according to another discussion (https://github.com/angular/angular/issues/34397) that is currently not possible.

**Comment by gkalpak** [comment]
I don't think that doing the work within `APP_INITIALIZER` solves the problem. You would still have to parse and execute the old JS and if an update is found you would still need to reload (although it is safer at that point, since the app has not bootstraped yet). In that regard, it is not much of an improvement over showing a spinner in the `AppComponent`.

##
Storing `index.html` in a data-group is not ideal, but it should work. Data-groups are geared towards data requests and therefore are treated slightly differently than assets. For example:
- They have an expiration date. (You can work-around it by setting a very high value.)
- They are not automatically downloaded (as [`prefetch` asset-groups][1] are), so `index.html` will not be cached until the second time it is requested. (Again, you could work around that by making an extra request to `index.html` once an update is installed.)

So, while this could work, it is a work-around and not recommended if #38565 covers your needs. BTW, you can grab the latest code (including the feature from #38565 from https://github.com/angular/service-worker-builds - for example `npm install --save https://github.com/angular/service-worker-builds`).

##
Regarding your option 4:
Yes you could do that, but as mentioned above it is far from ideal imo (i.e. I would personally prefer one of the other work-arounds).

Also, keep in mind that the SW does not immediately checks for updates on navigations requests. It _schedules_ a check for update to be run later (currently after 5 seconds). So, it is unlikely that the update will be available before the app bootstraps.

In any case, one modification I would suggest is to manually check for an update (via `SwUpdate#checkForUpdate()`). Note that if an update is found, it will be emitted on `SwUpdate#available` before the promise returned by `SwUpdate#checkForUpdate()` is resolved. So, by the time `SwUpdate#checkForUpdate()` resolves you will know whether there was an update or not.

[1]: https://angular.io/guide/service-worker-config#installmode

**Comment by romanovma** [comment]
> Also, keep in mind that the SW does not immediately checks for updates on navigations requests. It schedules a check for update to be run later (currently after 5 seconds). So, it is unlikely that the update will be available before the app bootstraps.

 Good to know. That actually gives us no choice but to use `index.html` in the `dataGroup` approach. Also, it allows setting a timeout for low network connections which is good and which is not part of the coming change in ver11 as I understood.

Thanks again for the detailed answers.
